compiler/frontend/src/mir_optimize/constant_lifting.rs

```diff
@@ -39,66 +39,60 @@ use rustc_hash::FxHashSet;
-                    .iter()
-                    .all(|captured| constant_ids.contains(captured));
-            if !is_constant {
-                return VisitorResult::Continue;
-            }
-            if is_return_value && let Expression::Reference(_) = expression {
-                // Returned references shouldn't be lifted. If we would lift
-                // one, we'd have to add a reference anyway.
-                return VisitorResult::Continue;
-            }
-            constants.push((id, expression.clone()));
-            constant_ids.insert(id);
-            VisitorResult::Continue
-        });
+    let mut index = 0;
+    while index < body.expressions.len() {
+        let (id, expression) = &body.expressions[index];
+        let id = *id;
+        let is_constant = expression.is_pure()
+            && expression
+                .captured_ids()
+                .iter()
+                .all(|captured| constant_ids.contains(captured));
+        let is_return_value = id == body.return_value();
 
-        self.body.visit_bodies(&mut |body| {
-            Self::remove_constants(body, &constant_ids, &mut self.id_generator)
-        });
-        Self::remove_constants(&mut self.body, &constant_ids, &mut self.id_generator);
-        for (_, expression) in &mut constants {
-            expression.visit_bodies(&mut |body| {
-                Self::remove_constants(body, &constant_ids, &mut self.id_generator);
-            })
+        if !is_constant {
+            index += 1;
+            continue;
+        }
+        if is_return_value && let Expression::Reference(_) = expression {
+            // Returned references shouldn't be lifted. If we would lift one,
+            // we'd have to add a reference anyway.
+            index += 1;
+            continue;
         }
 
-        self.body.insert_at_front(constants);
-    }
+        // This is a constant and should be lifted.
 
-    fn remove_constants(
-        body: &mut Body,
-        constant_ids: &FxHashSet<Id>,
-        id_generator: &mut IdGenerator<Id>,
-    ) {
-        let return_value = body.return_value();
-        body.remove_all(|id, _| constant_ids.contains(&id));
+        constants.push(body.expressions.remove(index));
+        constant_ids.insert(id);
 
-        if constant_ids.contains(&return_value) {
+        if is_return_value {
             // The return value was removed. Add a reference to the lifted
             // constant.
-            body.push(id_generator.generate(), Expression::Reference(return_value));
+            body.push(id_generator.generate(), Expression::Reference(id));
         }
     }
+
+    if constants.is_empty() {
+        return; // Nothing to lift.
+    }
+
+    let original_expression = mem::replace(expression, Expression::Parameter);
+    *expression = Expression::Multiple(Body::new(
+        constants
+            .into_iter()
+            .chain([(id_generator.generate(), original_expression)])
+            .collect_vec(),
+    ));
```

Comment:
```suggestion
    constants.push((id_generator.generate(), original_expression));
    *expression = Expression::Multiple(Body::new(constants));
```

Reply:
I thought about doing that, but the original expression isn't really a constant so it doesn't quite fit semantically. But I guess that's fine here.